Code intel strategy update (#2348)
sourcegraph/handbook: content/strategy-goals/strategy/code-graph/code-intelligence/index.md

### Vision

> We see Code intelligence as the glue that sticks the product together, providing a platform for features from navigation to precise powered searches, compiler-accurate batch changes and insights. We aim to provide support for all widely used languages and for the ones we don’t, provide a platform for any developer to add and test their own indexers.

> > In the near term we want Code intelligence to provide seamless, out-of-the-box, precise code navigation for languages that cover 90% of the market usage.

Reviewer: This is a great addition. All of these individual vision items need to be converted to start with `-` instead of `>` though because otherwise it all gets merged together in markdown (see image).

![image](https://user-images.githubusercontent.com/3529318/154218779-7f9aee45-ef0c-4015-9944-987df2b181c6.png)

> We want our code navigation to reach IDE feature parity, while offering the option of plugging into developers’ favorite IDEs.
> In the longer term, we envision building a global knowledge graph that accurately maps the entire code universe.

> 

Reviewer: In the guiding principles, the second one is great but I think the first one duplicates the Code Graph level principles so no need to repeat it here, unless you think it's specifically adding value.

> In the last few months we’ve:

Reviewer: I can't add a comment above, but I would change the sentence `Code intelligence lies at the very center of the product as a whole, providing the metadata that will eventually power other product areas.` to remove "will eventually power" and say "powers". This reality is already here.

Reviewer: I really like the maturity descriptions generally.

- Added precise support for the JVM ecosystem (Java, Scala and Kotlin).
- Shipped the [executors service](https://docs.sourcegraph.com/admin/executors) that powers [auto-indexing](https://docs.sourcegraph.com/code_intelligence/explanations/auto_indexing) at on-prem customers that use GCP or AWS.

Recent key learnings:

- We learned that LSIF manual setup is non-trivial and requires various levels of team involvement at enterprise customers. There’s an appetite for a out-of-the-box precise code navigation solution.
- Investing in platform performance and scalability is critical for the Global code graph vision to become a reality. Cloud is already at 2TB of data at ~40k indices. The global code graph for Java alone is somewhere between 800k to 6 million indices, so anywhere from 40TB to 300TB would be needed for the JVM code graph alone.
- Establishing a pricing strategy for the executors service proved to be non-trivial for all of our offerings.

> ### Top customer, support, sales and marketing issues

Reviewer 2: Is there a particular profile of customers (e.g. company size, industry, code base size, ...) that these requests come from? Or are we seeing these generally across the board? Does it make sense to try to contextualize these feedback/requests?

Reviewer: We should have a better process now for getting pricing decisions (still a recent learning, but just mentioning): https://handbook.sourcegraph.com/departments/product-engineering/product/process/#executive-decisions. 

https://docs.google.com/document/d/1Hj2AYA9RAb9zo7-cC4zoIrd68QRungyeGywCf20Sfwk/edit is meant to make this decision making flow even more transparent and clear, across all of leadership (still very much WIP, but that's where the work will go).

Author: @yinquanteo these are patterns generally across the board. I'll add a sentence to clarify and also link to product gaps (calling out they're internal only)

- Requests for adding precise support for more languages (often including the addition of cross-repository and cross-dependency navigation features).
- Current manual setup is not straightforward, customers run into issues while setting up LSIF for their repositories. This varies per indexer maturity and language ecosystem complexity.
- There have been reports about performance and scaling issues when indexing large monorepos.

> - Requests for a feature that helps visualize the code graph and its dependencies.

Reviewer 3: Requests for the ability to visualize...

> - Requests for a feature that helps visualize the code graph and its dependencies.

Author: ```suggestion
- Requests for the ability to visualize the code graph and its dependencies.
```

## Strategy and Plans

> ### Themes

Reviewer: This is a lot of themes to all be working on at once. Are some more important than others? Do you really expect to work on all of them actively now? In the summary below you touch on why they are all important and connected, but apart from that is it really possible to push all of them forward? If I look at what's next and why, it doesn't seem to address all these themes.

If some are things we want to get to later, they would go in what we're not working on and why (with the reason being that some of the other things are more urgent/important).

Author: This made me realize I was leaving out significant items from the What's next section. Thanks for the thoughtful catch!

> ### Themes

Reviewer 2: Are these scoped to FY23 Q1? @jyavorska, it might be worth calling that out in all our strategy pages so it's clear what the time frame is?

Reviewer: It's scoped to the strategy pages themselves, which are scoped to the next six months or a year or so. I'll create a separate PR to add something to the header of each strategy page that indicates that.

Reviewer 2: Sounds good, thanks!

**Unlock dependency navigation**

> We believe this is Code Intelligence's killer feature. Most developers think in terms of package and dependency versions rather than focusing on external repositories. Supporting precise dependency navigation elevates the code navigation experience to a new level of cross-project analysis. The ability to navigate to any third party dependency a repository references is key to use cases like [Developer onboarding](../use-cases/dev-onboarding.md) where time to value is directly related to how fast a developer is able to navigate and understand a new codebase (including the dependencies it relies on).

Author: ```suggestion
We believe this is Code Intelligence's killer feature. Most developers think in terms of package and dependency versions rather than focusing on external repositories. Supporting precise dependency navigation elevates the code navigation experience to a new level of cross-project analysis. The ability to navigate to any third party dependency a repository references is key to use cases like [Developer onboarding](../../use-cases/dev-onboarding.md) where time to value is directly related to how fast a developer is able to navigate and understand a new codebase (including the dependencies it relies on).
```

> We believe this is the global code graph’s killer feature. Most developers think in terms of package and dependency versions rather than focusing on external repositories. Supporting precise dependency navigation elevates the code navigation experience to a new level of cross-project analysis. The ability to navigate to any third party dependency a repository references is key to use cases like [Developer onboarding](../use-cases/dev-onboarding.md) where time to value is directly related to how fast a developer is able to navigate and undestand a new codebase (including the dependencies it relies on).

Reviewer 3: We believe this is Code Intelligence's killer feature?

Reviewer 4: ```suggestion
We believe this is Code Intelligence's killer feature. Most developers think in terms of package and dependency versions rather than focusing on external repositories. Supporting precise dependency navigation elevates the code navigation experience to a new level of cross-project analysis. The ability to navigate to any third party dependency a repository references is key to use cases like [Developer onboarding](../use-cases/dev-onboarding.md) where time to value is directly related to how fast a developer is able to navigate and understand a new codebase (including the dependencies it relies on).
```

> We believe this is the global code graph’s killer feature. Most developers think in terms of package and dependency versions rather than focusing on external repositories. Supporting precise dependency navigation elevates the code navigation experience to a new level of cross-project analysis. The ability to navigate to any third party dependency a repository references is key to use cases like [Developer onboarding](../use-cases/dev-onboarding.md) where time to value is directly related to how fast a developer is able to navigate and undestand a new codebase (including the dependencies it relies on).

Reviewer 3: I think it'd be good to alter this paragraph somewhat into how this dependency navigation coupled with cross-repository takes the IDE model and scales it to encompass the whole project/code/cognitive space of a company.

> We believe this is the global code graph’s killer feature. Most developers think in terms of package and dependency versions rather than focusing on external repositories. Supporting precise dependency navigation elevates the code navigation experience to a new level of cross-project analysis. The ability to navigate to any third party dependency a repository references is key to use cases like [Developer onboarding](../use-cases/dev-onboarding.md) where time to value is directly related to how fast a developer is able to navigate and undestand a new codebase (including the dependencies it relies on).

Author: ```suggestion
We believe this is the global code graph’s killer feature. Most developers think in terms of package and dependency versions rather than focusing on external repositories. Supporting precise dependency navigation elevates the code navigation experience to a new level of cross-project analysis. The ability to navigate to any third party dependency a repository references is key to use cases like [Developer onboarding](../../use-cases/dev-onboarding.md) where time to value is directly related to how fast a developer is able to navigate and undestand a new codebase (including the dependencies it relies on).
```

> We believe this is Code Intelligence's killer feature. Most developers think in terms of package and dependency versions rather than focusing on external repositories. Supporting precise dependency navigation elevates the code navigation experience to a new level of cross-project analysis. The ability to navigate to any third party dependency a repository references is key to use cases like [Developer onboarding](../../use-cases/dev-onboarding.md) where time to value is directly related to how fast a developer is able to navigate and understand a new codebase (including the dependencies it relies on).

Reviewer: I might say here that it is emerging as an additional killer feature. Existing code intelligence is also a killer feature. :)

Reviewer: This section is a really great addition and really explains why this is important. Maybe worth linking to Olaf's video? Is that internal only?

Author: Nice nitpick, I'll add it :)

Author: On Olaf's video, I couldn't remember why I hadn't added it before, I just rewatched and towards the end there's a screen share of one of our GH competition private docs. The loom is public though so I'm wondering if we should restrict it before adding a link to the strategy page.

Reviewer: Yeah you can either restrict it and link to it as (internal only), or plan to record another public video later. Both are fine.

> Taking it a step further, adding support for a wider set of artifactory hosts opens the door to a whole new array of cross-team features that support our use cases. These include (but are not limited to):

Reviewer 3: Is this mention of artifactory hosts referring to adding additional integration with package ecosystems such as PyPI?

Reviewer 4: Should probably be _package hosts_ then (or _artifact host_, but I think our term-of-art here is _package_).

> Taking it a step further, adding support for a wider set of artifactory hosts opens the door to a whole new array of cross-team features that support our use cases. These include (but are not limited to):

Author: ```suggestion
Taking it a step further, adding support for a wider set of package hosts opens the door to a whole new array of cross-team features that support our use cases. These include (but are not limited to):
```

> - Accelerating [Developer onboarding](../use-cases/dev-onboarding.md) by enabling search for dependency versions and across transitive dependencies.

Reviewer 3: "by enabling the ability to search over the full transitive dependency graph"?

> - Accelerating [Developer onboarding](../use-cases/dev-onboarding.md) by enabling search for dependency versions and across transitive dependencies.

Author: ```suggestion
- Accelerating [Developer onboarding](../../use-cases/dev-onboarding.md) by enabling the ability to search over the full transitive dependency graph.
```

> - Monitoring [Code Health](../use-cases/code-health.md) by creating dependency tracking insights to detect deprecated dependency usage.

Reviewer 3: Perhaps highlighting version drift is useful too?

> - Monitoring [Code Health](../use-cases/code-health.md) by creating dependency tracking insights to detect deprecated dependency usage.

Author: ```suggestion
- Monitor [Code Health](../../use-cases/code-health.md) by creating Code Insights that track deprecated dependency usages and version drift.
```

> - [Fixing security vulnerabilities](../use-cases/fixing-security-vulnerabilities.md) and helping [Incident response](../use-cases/incident-response.md) by searching and finding references to vulnerable dependency versions or errors, later applying a batch fix using Batch Changes.

Reviewer 3: I think the first point is good, having precise or search intel over *all* your code helps both those uses cases, precise being able to help on a much deeper level due to the ability to see the full transitive dependency graph. 

The latter "later applying a batch fix" is likely going to rely on precise intel no?  So perhaps we change it something along the lines of "being able to have an available fully semantic graph of the code in question would allow other products such as BC to do X"

> - [Fixing security vulnerabilities](../use-cases/fixing-security-vulnerabilities.md) and helping [Incident response](../use-cases/incident-response.md) by searching and finding references to vulnerable dependency versions or errors, later applying a batch fix using Batch Changes.

Author: ```suggestion
- [Fixing security vulnerabilities](../../use-cases/fixing-security-vulnerabilities.md) and helping [Incident response](../../use-cases/incident-response.md) by searching and finding references to vulnerable dependency versions or errors, later applying a batch fix using Batch Changes.
```

> - Encourage [Code Reuse](../use-cases/code-reuse.md) by making precise navigation a medium for finding references to relevant usage examples.

Reviewer 3: Perhaps I'm missing the audience here or I'm looking at the semantics of the sentence too much but I'd say find references (or code intel powered stuff like API Docs) would the *the* way, (not *a*) to power this use case.

Reviewer 4: ```suggestion
- Encourage [Code Reuse](../use-cases/code-reuse.md) by making precise navigation the obvious way to find relevant usage examples
```

> - Encourage [Code Reuse](../use-cases/code-reuse.md) by making precise navigation a medium for finding references to relevant usage examples.

Author: ```suggestion
- Encourage [Code Reuse](../../use-cases/code-reuse.md) by making precise navigation the obvious way to find references to relevant usage examples.
```

These are all currently in early stages of development and discussion. Read more about how we're thinking about the dependency graph in [RFC 593: Unlocking use cases based on the dependency graph of repositories](https://docs.google.com/document/d/1SkM8CG0IksvPEKRBRVLKipiRJTopx6Vq_hSWRJ9NyKs/edit#heading=h.trqab8y0kufp).

**Ship precise language support**

> We’ve historically invested in broadening our span of supported languages. This is an ongoing effort that ties directly back to the Global Code Graph vision. We're currently focusing on shipping a revamped version of our JS/TS indexer and building a Python indexer.

Reviewer: That it was our historical strategy is interesting, but can we explain here why we are choosing breadth over depth? Can be brief, but I think it makes sense to be clear about why.

Reviewer 2: Agreed with Jason. We could perhaps reference customer demand / product market fit?

Author: Added a reference to our language coverage goal tied to usage at customers/Cloud.

**Auto-indexing on-prem goes into Beta:**

> The current set up experience is not scalable for customers with a large amount of repositories. Enabling auto-indexing would mean a lower barrier for entry, a seamless experience and more engineers using precise code intelligence. Setting up auto-indexing is also a requirement for Sourcegraph to be able to index dependency packages.

Reviewer 3: Should we call out how strategically we're favouring accuracy over availability with precise (as in, to be precise we *have* to solve technical barriers that come with a cost) rather than availability over accuracy (going all in on a heuristic solution?).

> The current set up experience is not scalable for customers with a large amount of repositories. Enabling auto-indexing would mean a lower barrier for entry, a seamless experience and more engineers using precise code intelligence. Setting up auto-indexing is also a requirement for Sourcegraph to be able to index dependency packages.

Reviewer: Can we add something briefly on how it is not scalable? If it's just the manual setup steps, we can say that. Because otherwise the reader could interpret it in a lot of ways (i.e., is it unstable even when you have it set up? Does it lose accuracy at scale? etc.)

> The current set up experience is not scalable for customers with a large amount of repositories. Enabling auto-indexing would mean a lower barrier for entry, a seamless experience and more engineers using precise code intelligence. Setting up auto-indexing is also a requirement for Sourcegraph to be able to index dependency packages.

Reviewer 5: Would be cool to link to the auto-indexing docs page

Author: Done, thanks for catching!

Building the code graph also means we need to generate and store increased amounts of LSIF data that will require scaling our infrastructure in an order of one to two magnitudes. We hypothesize that we'll reach scaling concerns, we want to be proactive in identifying and removing bottlenecks.

Once we have validated our Experimental solution and have proven it’s running successfully at three customers, the next step is monitoring and weeding out any issues that might arise from our first trials to move on-prem auto-indexing into Beta. At this point we’ll be aiming to roll it out to a larger number of customers.

> **Make code navigation fast and reliable**

Reviewer: ```suggestion
**Make code navigation feel fast and reliable**
```

We receive a significant amount of feedback around papercuts and potential improvements to our navigation experience. We are commited to making our navigation feel fast and reliable to users. Our current focus is in solving low hanging navigation issues while also adding quality of life improvements (see [RFC 542](https://docs.google.com/document/d/1sDDpZaWdGtIaiNLNB8QsLwHTvH10fhEKpEa4qcog5vg/edit) for example). From a design perspective, we have plans to conduct research that helps us understand adoption drivers and pain points, with the aim of identifying concrete improvements to increase discoverability and enhance the navigation experience.

### How do these all tie together?

> Although the themes above might seem isolated or disconnected at a first glance, each one of them contributes to push forward our [Code Graph vision](../index.md). Auto-indexing serves a double purpose: it lowers the barrier to adoption by making precise code intelligence easier to configure, and also unlocks the ability to index dependencies (which play a central part in our current strategy). By adding more language support, combining it with adding artifactory hosts support, and improving our navigation features overall we're aiming to position Sourcegraph as a superior code navigation platform that goes beyond baseline expectations and creates delightful experiences for our users.

Reviewer 3: again, not sure artifactory is what you're wanting to refer to (see comment further up).

> Although the themes above might seem isolated or disconnected at a first glance, each one of them contributes to push forward our [Code Graph vision](../index.md). Auto-indexing serves a double purpose: it lowers the barrier to adoption by making precise code intelligence easier to configure, and also unlocks the ability to index dependencies (which play a central part in our current strategy). By adding more language support, combining it with adding artifactory hosts support, and improving our navigation features overall we're aiming to position Sourcegraph as a superior code navigation platform that goes beyond baseline expectations and creates delightful experiences for our users.

Reviewer 3: I think we could call out two themes here, one which you've done in that we think we can make a superior code navigation platform.  The second thing that we should add is similar to the sentiment here https://sourcegraph.slack.com/archives/C0LJP8WTB/p1644592718020329?thread_ts=1644582631.178299&cid=C0LJP8WTB 

That by providing this data layer we can unlock new features in our other product verticals (symbol renaming in batch changes, # of call sites to power ranking in search) or enhancing already existing product verticals (more accurate dependency tracking in code insights etc).  I think this is a key section of how Code Intelligence (or our other product verticals) may get matched by single offers in specific spaces and it's the intersection of all our verticals that provides the defensible market moat we want.

> Although the themes above might seem isolated or disconnected at a first glance, each one of them contributes to push forward our [Code Graph vision](../index.md). Auto-indexing serves a double purpose: it lowers the barrier to adoption by making precise code intelligence easier to configure, and also unlocks the ability to index dependencies (which play a central part in our current strategy). By adding more language support, combining it with adding artifactory hosts support, and improving our navigation features overall we're aiming to position Sourcegraph as a superior code navigation platform that goes beyond baseline expectations and creates delightful experiences for our users.

Author: ```suggestion
Although the themes above might seem isolated or disconnected at a first glance, each one of them contributes to push forward our [Code Graph vision](../index.md). Auto-indexing serves a double purpose: it lowers the barrier to adoption by making precise code intelligence easier to configure, and also unlocks the ability to index dependencies (which play a central part in our current strategy). By adding more language support, combining it with adding package hosts support, and improving our navigation features overall we're aiming to position Sourcegraph as a superior code navigation platform that goes beyond baseline expectations and creates delightful experiences for our users.
```

### What's next and why

> - **Adding precise C# and Ruby language support:** Based on our team's bandwidth and skill set, we are not planning to work on these languages this quarter. However, to support our [Grow ARR and Iterate with existing customers first](../../index.md#this-year-fy23) goals, we are working towards prioritizing these languages later this year based on current customer language share and associated ARR.

Reviewer: This ties into my comment above, but doing this next doesn't seem to address all the themes that you're saying you're actively driving.

Author: Adding extra items in what's next and why.

### What we're not working on & why

- **Incremental indexing for large monorepos:** When we ship auto-indexing for enterprise instances, we will likely need to solve incremental indexing to support our customers’ monorepos. This feature has been on our mid-term roadmap for quite some time now, but pain points have been worked through using workarounds like spacing LSIF upload frequencies depending on the customer’s repo size and commit frequency. At this point, incremental indexing could become a clear blocker and would be bumped up on our priority list.

- **Scale the C++ code graph:** Given the fragmented nature of the C/C++ environment we won’t be investing in improving C/C++ language support or scaling. We do however intend to revisit our solution in the future.

> - **Auto-indexing available for Cloud customers**

Reviewer: Great addition, it's so helpful to call these kinds of things out clearly.
